src/SparseMatrix.chpl

```diff
@@ -471,32 +471,79 @@ module SparseMatrix {
     import SymArrayDmap.makeSparseDomain;
     var (SD, dense) = makeSparseDomain(shape, layout);
 
-    for i in 0..<rows.size {
-      if SD.contains((rows[i], cols[i])) then
+    const minRow = min reduce rows;
+    const maxRow = max reduce rows;
+    const minCol = min reduce cols;
+    const maxCol = max reduce cols;
+
+    if minRow < 0 || minCol < 0 then
         throw getErrorWithContext(
-                                  msg="Duplicate index (%i, %i) in sparse matrix".format(rows[i], cols[i]),
-                                  lineNumber=getLineNumber(),
-                                  routineName=getRoutineName(),
-                                  moduleName=getModuleName(),
-                                  errorClass="InvalidArgumentError"
-                                  );
-      if rows[i] < 1 || rows[i] > shape[0] || cols[i] < 1 || cols[i] > shape[1] then
+                        msg="Sparse matrix indices must be greater than 0; got (%i, %i)".format(minRow, minCol), // TODO, change this when we start matrix from 0 instead of 1
+                        lineNumber=getLineNumber(),
+                        routineName=getRoutineName(),
+                        moduleName=getModuleName(),
+                        errorClass="InvalidArgumentError"
+                        );
+    if maxRow >= shape[0] || maxCol >= shape[1] then
+        throw getErrorWithContext(
+                        msg="Sparse matrix indices must be less than the shape; got (%i, %i) >= (%i, %i)".format(maxRow, maxCol, shape[0], shape[1]),
+                        lineNumber=getLineNumber(),
+                        routineName=getRoutineName(),
+                        moduleName=getModuleName(),
+                        errorClass="InvalidArgumentError"
+                        );
+
+    var A: [SD] eltType;
+    addElementsToSparseArray(A, SD, rows, cols, vals);
+
+    return A;
+  }
+
+  proc addElementsToSparseArray(ref A, ref SD, const ref rows, const ref cols, const ref vals) throws where A.chpl_isNonDistributedArray() {
+    for (r,c,v) in zip(rows, cols, vals) {
+      if A.domain.contains(r,c) then
         throw getErrorWithContext(
-                                  msg="Index (%i, %i) out of bounds for sparse matrix of shape (%i, %i)".format(rows[i], cols[i], shape[0], shape[1]),
+                  msg="Duplicate index (%i, %i) in sparse matrix".format(r, c),
+                  lineNumber=getLineNumber(),
+                  routineName=getRoutineName(),
+                  moduleName=getModuleName(),
+                  errorClass="InvalidArgumentError"
+                  );
+      SD += (r,c);
+      A[r,c] = v;
+    }
+  }
+
+
+  proc addElementsToSparseArray(ref A, ref SD, const ref rows, const ref cols, const ref vals) throws where !A.chpl_isNonDistributedArray() {
+    coforall (loc, locDom) in zip(getGrid(A),
+                                  SD._value.locDoms) {
```

Comment:
I am a big fan of 80-column coding and encourage you to use that (I'll push that agenda later on :D), but this line can be just made a single line given that the function header is already overlong.